src/components/structures/LeftPanel2.tsx

```diff
@@ -115,86 +115,130 @@ export default class LeftPanel2 extends React.Component<IProps, IState> {
-            });
-        } else {
+        window.requestAnimationFrame(() => {
             this.doStickyHeaders(list);
             this.isDoingStickyHeaders = false;
-        }
+        });
     }
 
     private doStickyHeaders(list: HTMLDivElement) {
-        const rlRect = list.getBoundingClientRect();
-        const bottom = rlRect.bottom;
-        const top = rlRect.top;
+        const topEdge = list.scrollTop;
+        const bottomEdge = list.offsetHeight + list.scrollTop;
         const sublists = list.querySelectorAll<HTMLDivElement>(".mx_RoomSublist2");
-        const headerRightMargin = 24; // calculated from margins and widths to align with non-sticky tiles
 
-        const headerStickyWidth = rlRect.width - headerRightMargin;
+        const headerRightMargin = 16; // calculated from margins and widths to align with non-sticky tiles
+        const headerStickyWidth = list.clientWidth - headerRightMargin;
+
+        // We track which styles we want on a target before making the changes to avoid
+        // excessive layout updates.
+        const targetStyles = new Map<HTMLDivElement, {
+            stickyTop?: boolean;
+            stickyBottom?: boolean;
+            makeInvisible?: boolean;
+        }>();
 
-        let gotBottom = false;
         let lastTopHeader;
+        let firstBottomHeader;
         for (const sublist of sublists) {
-            const slRect = sublist.getBoundingClientRect();
-
             const header = sublist.querySelector<HTMLDivElement>(".mx_RoomSublist2_stickable");
             header.style.removeProperty("display"); // always clear display:none first
 
-            if (slRect.top + HEADER_HEIGHT > bottom && !gotBottom) {
-                header.classList.add("mx_RoomSublist2_headerContainer_sticky");
-                header.classList.add("mx_RoomSublist2_headerContainer_stickyBottom");
-                header.style.width = `${headerStickyWidth}px`;
-                header.style.removeProperty("top");
-                gotBottom = true;
-            } else if (((slRect.top - (HEADER_HEIGHT * 0.6) + HEADER_HEIGHT) < top) || sublist === sublists[0]) {
-                // the header should become sticky once it is 60% or less out of view at the top.
-                // We also add HEADER_HEIGHT because the sticky header is put above the scrollable area,
-                // into the padding of .mx_LeftPanel2_roomListWrapper,
-                // by subtracting HEADER_HEIGHT from the top below.
-                // We also always try to make the first sublist header sticky.
-                header.classList.add("mx_RoomSublist2_headerContainer_sticky");
-                header.classList.add("mx_RoomSublist2_headerContainer_stickyTop");
-                header.style.width = `${headerStickyWidth}px`;
-                header.style.top = `${rlRect.top - HEADER_HEIGHT}px`;
+            // When an element is <=40% off screen, make it take over
+            const offScreenFactor = 0.4;
```

Comment:
@bwindels this number is fairly arbitrary, but should match the old 60% value with the older math. Might need some tweaking on the refresh.